CONTRIBUTING.md

```diff
@@ -71,6 +75,72 @@ IntelliJ's code generation can be configured to do this automatically:
      Tick:
      - Make generated local variables final
      - Make generated parameters final
+     
+**Null checking of method or function parameters**
+
+If arguments to a constructor set members of the object, or if arguments to a method
```

Comment:
Null parameter checking seems to have come over from C++ programming. In C/C++ if you dereference an invalid pointer the results are undefined - you might get a segfault, but your program might continue, now silently referencing a garbage object, potentially causing all sorts of nasty knock-on effects.
But in Java that can't happen - the only kind of invalid pointer that's possible is a null pointer, and an attempt to dereference it is guaranteed to throw a NullPointerException. When this happens (hopefully when running your test suite!) you look at the stack trace and fix the problem. Adding an Objects.requireNotNull() also throws a NullPointerException.
So you're just trading one NPE for another, either of which will allow you to solve the issue. So you're just adding extra unsightly code for no actual gain. Without the check you will get the NPE whenever there's an attempt to dereference, so it's going to happen in your test suite run if this happens by any execution path. So assuming you have written tests it's going to fire. If you haven't written tests you have a bigger issue to solve first! The fact that it will fire in your test suite makes the argument "I want to throw early" kind of meaningless.
There is one exception to this where i would recommend param checking - and that's in public APIs - here you want to throw exceptions on invalid params immediately as you don't want to expose a stack trace from internal implementation classes through your API.
I recommend we drop this rule

Reply:
I'm a big fan of not letting objects get into inconsistent states.  This actually goes beyond just null checking references in constructors. It includes any method where a parameter passed in is stored within the state of the object and it includes not just null checking but any other parameter state that may put the instance into an invalid state, e.g. a negative count, an empty string, etc.

I'm of fan of this because:
1. Useful stack trace: Most importantly, you get the stack trace of where the error occurred. If you remove the check you get a stack trace of where the bad state is accessed.  This can be in a totally different part of the code and at a much later time.  I've had the pleasure of trying to track down these kinds of bugs many times and they are a nightmare. 
1. It results in cleaner code.  If you know a field is never null, or a long is never negative, then the code accessing those fields doesn't need any error checking.  Without such checking its common to see return values from getters having error checking applied and such error checking is spread across the code base.  You can of course suggest that such error checking is not necessary, but that doesn't stop people adding it.
1. It makes the code explicit, easier to grok and to reason about.  It's clear to anyone reading the code that this field is never going to be null.

Reply:
> I'm a big fan of not letting objects get into inconsistent states. This actually goes beyond just null checking but any other parameter state that may put the instance into an invalid state, e.g. a negative count, an empty string, etc. It's good defensive programming.

The point is specifically about null pointer checking, not about parameter checking in general.

> 
> I'm of fan of this because:
> 
> 1. Useful stack trace: Most importantly, you get the stack trace of where the error occurred, i.e. the source of where the bad data/state is coming from. If you remove the check, you get a stack trace of where the bad state is accessed.  This can be in a totally different part of the code and at a much later time.  I've had the pleasure of trying to track down these kinds of bugs many times and they are a nightmare.

I don't follow this logic. If you don't have the null pointer check then you will get a NPE whenever that member is accessed - that will be during the same test run as if you'd added an explicit null pointer check, so it won't be "later" on in any meaningful way. The stack trace will show the class and the member that is null. You then just need to put a breakpoint on the constructor of the class where that member is set and run the failing test. You will then see the exact same stacktrace as you'd get if you'd put an explicit null pointer check in. 

> 2. It makes the code explicit, easier to grok and to reason about.  It's clear to anyone reading the code that this field is never going to be null.

We should assume that all object member variables should never be null. We should use optional to denote something that may or may not be set. So I don't think null pointer checks add anything here.

> 3. It results in cleaner code.  If you know a field is never null,

As above. No object fields should ever be null. I'd argue that it makes the code much less clean due to the extra visual clutter.

>  or a long is never negative,

The point is specifically about null pointer checking, not param checking in general. Sure, other types of param checking might be valid, but generally it's better to do those on the boundaries of the system, e.g. validate config at the edges before it gets passed into the main body of the code.

Unlike C++ Java already provides a built in mechanism for checking whether an object reference is null - it's called throwing NPE! ;)

Reply:
https://openjdk.java.net/jeps/358 gives excellent info on NPEs thrown by the JVM

Reply:
> I don't follow this logic. If you don't have the null pointer check then you will get a NPE whenever that member is accessed - that will be during the same test run as if you'd added an explicit null pointer check, so it won't be "later" on in any meaningful way. The stack trace will show the class and the member that is null. You then just need to put a breakpoint on the constructor of the class where that member is set and run the failing test. You will then see the exact same stacktrace as you'd get if you'd put an explicit null pointer check in.

There's always going to be cases we miss and hit in the wild. These also tend to be the hardest to debug (since they are not always easy to reproduce with a test case). Having a stack trace that tells us what wrote the null value is helpful when that happens.

Reply:
> The point is specifically about null pointer checking, not about parameter checking in general.

I'm not sure why you're proposing we treat null pointer checking differently from any other parameter validation. It's all the same thing: defensive programming.  If you're going to store a parameter in the state of your object, you should ensure its valid so that you object does not get into an invalid state.

> I don't follow this logic. If you don't have the null pointer check then you will get a NPE whenever that member is accessed - that will be during the same test run as if you'd added an explicit null pointer check, so it won't be "later" on in any meaningful way.

If all bugs were found by tests and fixed, then I'd be in agreement with you that null checking, or any parameter validation, is unnecessary. However, there are always bugs that get missed and are only found in production. Given the choice between a stack trace showing what was trying to put an object into an invalid state vs one that from when the invalid state is accessed, I know which I'd prefer.